Applications/Spire/Source/KeyBindings/TaskKeysPage.cpp

```diff
@@ -73,18 +87,155 @@ TaskKeysPage::TaskKeysPage(std::shared_ptr<KeyBindingsModel> key_bindings,
 }
 
 const std::shared_ptr<KeyBindingsModel>&
     TaskKeysPage::get_key_bindings() const {
   return m_key_bindings;
 }
+
+bool TaskKeysPage::eventFilter(QObject* watched, QEvent* event) {
+  if(watched == m_table_view) {
+    if(event->type() == QEvent::KeyPress) {
+      auto& key_event = *static_cast<QKeyEvent*>(event);
+      if(key_event.modifiers() & Qt::ShiftModifier &&
+          (key_event.key() == Qt::Key_Return ||
+            key_event.key() == Qt::Key_Enter)) {
+        on_new_task_action();
+        return true;
+      } else if(key_event.modifiers() & Qt::ControlModifier &&
+          key_event.key() == Qt::Key_D) {
+        on_duplicate_task_action();
+        return true;
+      }
+    }
+  } else if(watched == m_added_region_item) {
+    if(event->type() == QEvent::Show) {
+      auto enter_event =
+        QKeyEvent(QEvent::KeyPress, Qt::Key_Enter, Qt::NoModifier);
+      QCoreApplication::sendEvent(&m_added_region_item->get_body(),
+        &enter_event);
+      m_added_region_item->removeEventFilter(this);
+    }
+  }
+  return QWidget::eventFilter(watched, event);
+}
+
+void TaskKeysPage::update_button_state() {
+  auto is_enabled =
+    m_table_view->get_selection()->get_row_selection()->get_size() > 0;
+  m_duplicate_button->setEnabled(is_enabled);
+  m_delete_button->setEnabled(is_enabled);
+}
+
+void TaskKeysPage::on_new_task_action() {
+  m_new_task_form->show();
+  auto window = m_new_task_form->window();
+  window->move(
+    mapToGlobal(QPoint(0, 0)) + rect().center() - window->rect().center());
+}
+
+void TaskKeysPage::on_duplicate_task_action() {
+  auto& selection = m_table_view->get_selection()->get_row_selection();
+  if(selection->get_size() == 0) {
+    return;
+  }
+  m_table_view->setFocus();
+  auto last_current_row = m_table_view->get_current()->get()->m_row;
+  auto sorted_selection = std::vector<int>();
+  sorted_selection.assign(selection->begin(), selection->end());
+  std::sort(sorted_selection.begin(), sorted_selection.end(), std::greater<int>());
+  auto& table = m_table_view->get_body().get_table();
+  for(auto index : sorted_selection) {
+    auto order_task = OrderTaskArguments();
+    order_task.m_name = table->get<QString>(index, 1);
+    order_task.m_region = table->get<Region>(index, 2);
+    order_task.m_destination = table->get<Destination>(index, 3);
+    order_task.m_order_type = table->get<OrderType>(index, 4);
+    order_task.m_side = table->get<Side>(index, 5);
+    order_task.m_quantity = table->get<optional<Quantity>>(index, 6);
+    order_task.m_time_in_force = table->get<TimeInForce>(index, 7);
+    order_task.m_additional_tags =
+      table->get<std::vector<Nexus::Tag>>(index, 8);
+    m_key_bindings->get_order_task_arguments()->insert(order_task,
+      m_table_view->get_current()->get()->m_row);
+  }
+  QTimer::singleShot(0, this, [=] {
+    m_table_view->get_current()->set(TableView::Index(last_current_row, 1));
```

Comment:
Why is this done after the current event loop?

Reply:
After copying rows, the current cell should be `row_selection[0][1]` based on specs. When new rows are inserted to the table, the current row will remain unchanged instead of the newly added row.